File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/ObjectWithCompositeId.java
package io.quarkus.it.panache.reactive;

import java.io.Serializable;
import java.util.Objects;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.IdClass;

import io.quarkus.hibernate.reactive.panache.PanacheEntityBase;

@Entity
@IdClass(ObjectWithCompositeId.ObjectKey.class)
public class ObjectWithCompositeId extends PanacheEntityBase {
    @Id
    public String part1;
    @Id
    public String part2;
    public String description;

    static class ObjectKey implements Serializable {
        private String part1;
        private String part2;

        public ObjectKey() {
        }

        public ObjectKey(String part1, String part2) {
            this.part1 = part1;
            this.part2 = part2;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o)
                return true;
            if (o == null || getClass() != o.getClass())
                return false;
            ObjectKey objectKey = (ObjectKey) o;
            return part1.equals(objectKey.part1) &&
                    part2.equals(objectKey.part2);
        }

        @Override
        public int hashCode() {
            return Objects.hash(part1, part2);
        }
    }
}

File: integration-tests/hibernate-reactive-panache-kotlin/src/main/kotlin/io/quarkus/it/panache/reactive/kotlin/ObjectWithCompositeId.kt
package io.quarkus.it.panache.reactive.kotlin

import io.quarkus.hibernate.reactive.panache.kotlin.PanacheCompanionBase
import io.quarkus.hibernate.reactive.panache.kotlin.PanacheEntityBase
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.IdClass
import java.io.Serializable

@Entity
@IdClass(ObjectWithCompositeId.ObjectKey::class)
class ObjectWithCompositeId : PanacheEntityBase {
    companion object : PanacheCompanionBase<ObjectWithCompositeId, ObjectKey>

    @Id
    var part1: String? = null

    @Id
    var part2: String? = null
    var description: String? = null

    internal class ObjectKey : Serializable {
        private var part1: String? = null
        private var part2: String? = null

        constructor() {}
        constructor(part1: String?, part2: String?) {
            this.part1 = part1
            this.part2 = part2
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ObjectKey) return false

            if (part1 != other.part1) return false
            if (part2 != other.part2) return false

            return true
        }

        override fun hashCode(): Int {
            var result = part1?.hashCode() ?: 0
            result = 31 * result + (part2?.hashCode() ?: 0)
            return result
        }
    }
}
